Tests/ParseSwiftTests/ParseObjectBatchTests.swift

```diff
@@ -136,7 +136,7 @@ class ParseObjectBatchTests: XCTestCase { // swiftlint:disable:this type_body_le
             switch saved[0] {
 
             case .success(let first):
-                XCTAssertEqual(first, scoreOnServer)
+                XCTAssert(first.hasSameObjectId(as: scoreOnServer))
```

Comment:
I'm not sure what happened here? The ParseObjects had something that already compared the objects directly by using ==. I know there was a discussion with @mtrezza about how the other SDKs handled it. It seems like we shouldn't require the developers to call `hasSameObjectId` and should do it for them as one may naturally attempt to compare two objects using ==

Reply:
Maybe there's a misunderstanding about that discussion. My thought was that we should leave equality and this equality-esque thing separate, and my goal was to leave `==` for cases where we want to check strict, deep, object equality, if that ever comes.

Reply:
Are you saying that == is still there? I assumed it was removed because these test cases were changed  from == to use `hasSameObjectId`. Did you only do this on a few to see if hasSameObjectId was working and left the rest of the == comparisons?

Reply:
No, there's currently no implementation for `==`. I replaced all the test cases that were doing equality with `==` with `hasSameObjectId`, because the implementation for `hasSameObjectId` is doing what the old `==` implementation did. At some point in the future, we'll have to provide a new implementation for `==` that does a real equality check.

Reply:
Hmm, I’m not sure I agree with this. The objects with the same objectId should be == imo. Now if one or many of those object are dirty, they are still the same object, but “may” have different/outdated data. At the end of the day, that should get resolved at the parse-server level and depending on the use-case, the developer will add Cloud Code to resolve. When the resolution/synchronization occurs at the server level, all objects will be hydrated back with the new/same data (assuming there’s an Internet connection of course). So objects with the same objectId are the same, rather they are in sync is a different question. This goes into the data store convo we were originally having with @drdaz

Reply:
The sync comments are above/below here: https://github.com/parse-community/Parse-Swift/issues/3#issuecomment-657700606

Reply:
The ObjC SDK detects dirty keys locally, FWIW. In either case, we can go the direction you're proposing, but I want to point out the following line from the docs for `Swift.Equatable`:

> Equality implies substitutability—any two instances that compare equally can be used interchangeably in any code that depends on their values.

If we don't adhere to this, we're not following the conventions for Equatable, and that might have unintended consequences down the road.

Reply:
I believe the parts you mention speak to the points I was making about the same objectId, they are definitely interchangeable because they have the same objectId. Once a sync happens they should all have the same data, meaning they are ==.

Reply:
Right, so objects with the same data should compare equal using `==`, right? And objects with the same ID can be compared directly using their IDs. Another thing I just thought of is that if we override `==`, users will no longer be able to rely on Swift to synthesize an implementation for deep equality.

Reply:
I'm saying it's harder to compare if two objects are actually the same and I don't think that should be the goal of ==. == should compare the same objectId because when they resolve (typically at the pare-server level) they are exactly the same. There's no way to guarantee all of the data is the same when all the data is local at the device and we shouldn't be making any promises here because of synchronization. I wouldn't try to compare the data locally, if there's a synching problem, usually apps ask the user which data do they want to keep without making a decision for them. In a centralized case, the parse-server will be asking for this resolution

Reply:
I'll say the big issue here is because our objects are not immutable (I'm not arguing for or against), if they were, a new objectId will be given to an object whenever a modification occurs, requiring a connection to the parse-server to get the new objectId of course. In this way, you would ensure that two objects with the same objectId actually have the same data.

Reply:
That makes sense, but how would we implement deep equality in the Swift SDK if we replace `==`? We'd squash the implementation that the compiler synthesizes it. I just want to hear what the *advantage* of having `==` compare IDs would be.

The way I see it, there isn't one: we (and users) can very easily fall back to `hasSameObjectId`, *and* we aren't breaking the definition of Equatable. I don't think 2 objects with the same ID are equivalent if they have different data. For example, look at this situation:

```swift
class MyViewController: UIViewController {
  var gameScore: GameScore = ... // gets set on init or something

  func configureUI(with gameScore: GameScore) {
    guard gameScore != self.gameScore else { return } // to avoid repeating the same work
    // update the UI
  }
}
```

If `==` compares IDs, this (and countless other situations where users are trying to check real equality) are impossible. As for making promises about synchronization, without a way to inspect dirty keys, we are definitely lacking functionality that's currently present in the ObjC and Android SDKs, which is only sending back dirty keys to a server when an object is updated. The idea is to conserve bandwidth: if you download an object with 1000 fields and update 2 of them locally, the device only sends the 2 that have been updated back to the server. We certainly cannot guarantee that the local copy of an object we have is the same as the copy on the server, but we can and should be able to guarantee that 2 local copies do or don't have the same data.

I think it's clear at this point which approach I prefer, but I'm not unwilling to consider having `==` compare IDs; **I'd just like to know what material and concrete advantages that would provide.** There's a real concrete advantage in NOT doing anything with `==`, which is getting the compiler implementation of Equatable for free; then, there are several subjective advantages like:
* Sticking to the definition of Equatable.
* Following the conventions that every Swift struct already has; users expect their structs to compare equal using `==` if they have the same data.
* Sticking to the [Principle of Least Astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Another approach would be to leave `Equatable` out of the SDK entirely and just let users decide what they want to do for their own use case. Internally, we could use `hasSameObjectId` and a similar helper for deep equality if we need it. I'm open to any of the 3 solutions, but I'd like to do the one that is the most beneficial and the least surprising for end users.

Reply:
`hasSameObject` can work and maybe it’s okay for this to leave Equatable in tact, so I’m fine for now

For the situation you mentioned, I would expect the developer to pull the latest version of game score from the Parse-Server before displaying, refreshing self.gamescore. If they don’t have access to Internet, they may hit their cache (or local storage) for their latest local version. Using the regular wall clock scenario, they would determine the latest by sorting by updatedAt, but all of this is depending on use-case and setup...

Reply:
I think our best bet is to revisit later based on feedback and what we think when we use it in apps in real life. For now, we can stick with what's there, and adapt as we move on.